leader/leader.go

```diff
@@ -175,51 +185,51 @@ func (l *Client) AddWatch(key string, retry time.Duration, valuesC chan string)
 				}
 			}
 
-			if err != nil {
-				select {
-				case <-ticker.C:
-					steps += 1
+			// The watcher has received a new event, send it to the client
+			// unless this value has already been sent.
+			if err == nil {
+				if resp.PrevNode != nil && resp.PrevNode.Value == resp.Node.Value && sentAnything {
+					continue
 				}
-
-				if err == context.Canceled {
+				select {
+				case valuesC <- resp.Node.Value:
+					sentAnything = true
+				case <-l.closeC:
+					logger.Info("Watcher is closing.")
 					return
-				} else if cerr, ok := err.(*client.ClusterError); ok {
-					if len(cerr.Errors) != 0 && cerr.Errors[0] == context.Canceled {
-						return
-					}
-					log.Debugf("unexpected cluster error: %v (%v)", err, cerr.Detail())
-					continue
-				} else if IsWatchExpired(err) {
-					log.Debug("watch expired, resetting watch index")
-					watcher, resp, err = l.getWatchAtLatestIndex(ctx, api, key, retry)
-					if err != nil {
-						continue
-					}
-				} else {
-					log.Warningf("unexpected watch error: %v", err)
-					// try recreating the watch if we get repeated unknown errors
-					if steps > 10 {
-						watcher, resp, err = l.getWatchAtLatestIndex(ctx, api, key, retry)
-						if err != nil {
-							continue
-						}
-						backoff.Reset()
-						steps = 0
-					}
-
-					continue
 				}
-			}
-			// if nothing has changed and we previously sent this subscriber this value,
-			// do not bother subscriber with extra notifications
-			if resp.PrevNode != nil && resp.PrevNode.Value == resp.Node.Value && sentAnything {
 				continue
 			}
+
+			// Otherwise, the watcher encountered an error while waiting for
+			// next event, we may need to recreate it.
 			select {
-			case valuesC <- resp.Node.Value:
-				sentAnything = true
-			case <-l.closeC:
+			case <-ticker.C:
+				steps += 1
+			}
+
+			if IsContextCanceled(err) {
+				// The context has been canceled while watcher was waiting
+				// for next event.
+				logger.Info("Context has been canceled.")
 				return
+			} else if clusterErr, ok := err.(*client.ClusterError); ok {
+				// Got an etcd error, the watcher will retry.
+				logger.WithError(clusterErr).Errorf("Etcd error: %v.", clusterErr.Detail())
+			} else if IsWatchExpired(err) {
+				// The watcher has expired, reset it so it's recreated on the
+				// next loop cycle.
+				logger.Info("Watch has expired, resetting watch index.")
+				watcher = nil
+			} else {
+				// Unknown error, try recreating the watch after a few repeated
+				// unknown errors.
+				logger.WithError(err).Error("Unexpected watch error.")
+				if steps > 10 {
+					watcher = nil
+					backoff.Reset()
+					steps = 0
+				}
```

Comment:
Overall feeling of the function is that it's still complicated to read. Can we move it to outside of `AddWatch` and maybe replace comments with code?
Something along the lines of (would be more readable if the loop/tick/classify were implemented by a custom type):
```go
func (l *Client) watchLoop(ctx context.Context, api client.KeysAPI, key string, retry time.Duration, valuesC chan string, logger logrus.FieldLogger) {
	var watcher client.Watcher
	var resp *client.Response
	var err error
	backoff := NewUnlimitedExponentialBackOff()
	ticker := ebackoff.NewTicker(backoff)
	var steps int
	// classify errors and decide what to do.
	// returns false to abort and exit or true to continue the loop
	classify := func(err error) (ok bool) {
		if IsContextCanceled(err) {
			// The context has been canceled while watcher was waiting
			// for next event.
			logger.Info("Context has been canceled.")
			return false
		}
		if clusterErr, ok := err.(*client.ClusterError); ok {
			// Got an etcd error, the watcher will retry.
			logger.WithError(clusterErr).Errorf("Etcd error: %v.", clusterErr.Detail())
		} else if IsWatchExpired(err) {
			// The watcher has expired, reset it so it's recreated on the
			// next loop cycle.
			logger.Info("Watch has expired, resetting watch index.")
			watcher = nil
		} else {
			// Unknown error, try recreating the watch after a few repeated
			// unknown errors.
			logger.WithError(err).Error("Unexpected watch error.")
			if steps > 10 {
				watcher = nil
				backoff.Reset()
				steps = 0
			}
		}
		return true
	}
	tick := func(err error) (ok bool) {
		select {
		case <-l.closeC:
			return false
		case <-ticker.C:
			steps += 1
			return classify(err)
		}
	}
	var sentAnything bool

	for tick(err) {
		watcher, err = l.getWatchAtLatestIndex(ctx, api, key, retry, valuesC)
		if err != nil {
			continue
		}

		resp, err = watcher.Next(ctx)
		if err != nil {
			continue
		}

		if resp.Node.Value == "" {
			continue
		}

		if resp.PrevNode != nil && resp.PrevNode.Value == resp.Node.Value && sentAnything {
			continue
		}

		select {
		case valuesC <- resp.Node.Value:
			sentAnything = true
		case <-l.closeC:
			logger.Info("Watcher is closing.")
			return
		}
	}
}
```